Fix NJT linear_backward() memory usage

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141163

Fixes #141112

The formula we're using for `linear_backward()` is inefficient for higher dim input sizes, even if the input is trivially higher dim (e.g. via use of `unsqueeze()`). This PR updates the formula to match the more efficient version employed by NST. Specifically, note the leading dim collapse for `grad_output`'s values before we compute the various matmuls.
https://github.com/pytorch/pytorch/blob/d5ee1d1b581da8399d604bd661ea5fe454b485d6/aten/src/ATen/native/nested/NestedTensorBackward.cpp#L37-L70

Testing for correctness is done via existing gradcheck tests (e.g. `test_backward_nn_functional_linear`). I added a memory usage test but I think it's likely there's a better way to do this.

jbschlosser committed:
Fix NJT linear_backward() memory usage

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix NJT linear_backward() memory usage"




[ghstack-poisoned]

jbschlosser committed:
Update on "Fix NJT linear_backward() memory usage"


Fixes #141112

The formula we're using for `linear_backward()` is inefficient for higher dim input sizes, even if the input is trivially higher dim (e.g. via use of `unsqueeze()`). This PR updates the formula to match the more efficient version employed by NST. Specifically, note the leading dim collapse for `grad_output`'s values before we compute the various matmuls.
https://github.com/pytorch/pytorch/blob/d5ee1d1b581da8399d604bd661ea5fe454b485d6/aten/src/ATen/native/nested/NestedTensorBackward.cpp#L37-L70

Testing for correctness is done via existing gradcheck tests (e.g. `test_backward_nn_functional_linear`). I added a memory usage test but I think it's likely there's a better way to do this.

[ghstack-poisoned]

jbschlosser commented:
Discussed offline: reset the max memory stat via `torch.cuda.reset_max_memory_allocated()` and measure max afterwards. If it's too high (in practice, I see over 3 GB allocated from the backward call), fail the test. Assuming this stat is process-isolated, this should work fine (we don't run tests in CI multi-threaded, only multi-process). If the test fails later on, we can revisit this but at least the fix is in :)

jbschlosser commented:
@pytorchbot merge

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -544,14 +544,16 @@ def linear_backward_default(func, *args, **kwargs):
         )
     if output_mask[1]:
-        dw = torch.matmul(grad_output._values.transpose(-2, -1), inp._values)
+        dw = torch.matmul(reshaped_grad.t(), inp._values.reshape(-1, weight.size(1)))
```
Might be worth a small comment about the optimization we're doing here :)

jbschlosser replied:
agreed, adding a desc in the next iteration

jbschlosser committed:
Update on "Fix NJT linear_backward() memory usage"


Fixes #141112

The formula we're using for `linear_backward()` is inefficient for higher dim input sizes, even if the input is trivially higher dim (e.g. via use of `unsqueeze()`). This PR updates the formula to match the more efficient version employed by NST. Specifically, note the leading dim collapse for `grad_output`'s values before we compute the various matmuls.
https://github.com/pytorch/pytorch/blob/d5ee1d1b581da8399d604bd661ea5fe454b485d6/aten/src/ATen/native/nested/NestedTensorBackward.cpp#L37-L70

Testing for correctness is done via existing gradcheck tests (e.g. `test_backward_nn_functional_linear`). I added a memory usage test but I think it's likely there's a better way to do this.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
